File: framework/backend/kotlin/packages/k1s0-consensus/src/main/kotlin/dev/k1s0/consensus/SagaOrchestrator.kt
package dev.k1s0.consensus

import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.Dispatchers
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.jetbrains.exposed.sql.Database
import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
import java.time.Instant
import java.util.UUID
import kotlin.time.Duration.Companion.milliseconds

private val logger = KotlinLogging.logger {}

/**
 * Orchestrator-based saga execution engine with PostgreSQL persistence.
 *
 * Executes saga steps sequentially. On failure, runs compensation steps
 * in reverse order. Failed sagas are retried according to the retry policy.
 * After exhausting retries, sagas are moved to the dead letter queue.
 *
 * @property database The Exposed database instance.
 * @property sagaConfig Saga configuration.
 * @property metrics Optional saga metrics.
 */
public class SagaOrchestrator(
    private val database: Database,
    private val sagaConfig: SagaConfig = SagaConfig(),
    private val metrics: SagaMetrics? = null,
) {
    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Executes a saga definition.
     *
     * @param definition The saga definition to execute.
     * @param initialContext Optional initial context data.
     * @return The result of the saga execution.
     */
    public suspend fun execute(
        definition: SagaDefinition,
        initialContext: Map<String, Any> = emptyMap(),
    ): SagaResult {
        val sagaId = UUID.randomUUID().toString()
        val context = initialContext.toMutableMap()
        val completedSteps = mutableListOf<String>()

        logger.info { "Starting saga '${definition.name}' (id=$sagaId)" }
        metrics?.sagaStarted(definition.name)

        persistSagaInstance(sagaId, definition.name, SagaStatus.RUNNING, 0, context)

        for ((index, step) in definition.steps.withIndex()) {
            try {
                logger.debug { "Executing step '${step.name}' in saga '$sagaId'" }
                updateSagaStep(sagaId, index, SagaStatus.RUNNING)
                step.execute(context)
                completedSteps.add(step.name)
                logger.debug { "Step '${step.name}' completed in saga '$sagaId'" }
            } catch (e: Exception) {
                logger.error(e) { "Step '${step.name}' failed in saga '$sagaId'" }
                metrics?.sagaStepFailed(definition.name, step.name)

                val compensatedSteps = compensate(sagaId, definition, completedSteps, context)

                val retryCount = getRetryCount(sagaId)
                if (retryCount < definition.retryPolicy.maxRetries) {
                    updateSagaRetry(sagaId, retryCount + 1, e.message)
                    logger.info { "Saga '$sagaId' will be retried (attempt ${retryCount + 1}/${definition.retryPolicy.maxRetries})" }
                } else if (sagaConfig.deadLetterEnabled) {
                    moveToDeadLetter(sagaId, e.message)
                    metrics?.sagaDeadLettered(definition.name)
                    logger.warn { "Saga '$sagaId' moved to dead letter queue after ${definition.retryPolicy.maxRetries} retries" }
                    return SagaResult(
                        sagaId = sagaId,
                        status = SagaStatus.DEAD_LETTER,
                        context = context,
                        error = e.message,
                        completedSteps = completedSteps,
                        compensatedSteps = compensatedSteps,
                    )
                }

                return SagaResult(
                    sagaId = sagaId,
                    status = if (compensatedSteps.isNotEmpty()) SagaStatus.COMPENSATED else SagaStatus.FAILED,
                    context = context,
                    error = e.message,
                    completedSteps = completedSteps,
                    compensatedSteps = compensatedSteps,
                )
            }
        }

        updateSagaStatus(sagaId, SagaStatus.COMPLETED)
        metrics?.sagaCompleted(definition.name)
        logger.info { "Saga '${definition.name}' completed (id=$sagaId)" }

        return SagaResult(
            sagaId = sagaId,
            status = SagaStatus.COMPLETED,
            context = context,
            completedSteps = completedSteps,
        )
    }

    /**
     * Resumes a saga from its persisted state.
     *
     * @param sagaId The ID of the saga to resume.
     * @param definition The saga definition.
     * @return The result of the resumed saga execution.
     */
    public suspend fun resume(sagaId: String, definition: SagaDefinition): SagaResult? {
        val instance = getSagaInstance(sagaId) ?: return null

        if (instance.status == SagaStatus.COMPLETED || instance.status == SagaStatus.DEAD_LETTER) {
            logger.info { "Saga '$sagaId' is already in terminal state: ${instance.status}" }
            return SagaResult(
                sagaId = sagaId,
                status = instance.status,
                context = emptyMap(),
                error = instance.error,
            )
        }

        logger.info { "Resuming saga '$sagaId' from step ${instance.currentStep}" }
        val context = mutableMapOf<String, Any>()
        val completedSteps = mutableListOf<String>()

        // Skip already completed steps
        for (i in 0 until instance.currentStep) {
            completedSteps.add(definition.steps[i].name)
        }

        // Continue from the failed step
        for (i in instance.currentStep until definition.steps.size) {
            val step = definition.steps[i]
            try {
                step.execute(context)
                completedSteps.add(step.name)
            } catch (e: Exception) {
                logger.error(e) { "Resumed step '${step.name}' failed in saga '$sagaId'" }
                val compensated = compensate(sagaId, definition, completedSteps, context)
                return SagaResult(
                    sagaId = sagaId,
                    status = SagaStatus.COMPENSATED,
                    context = context,
                    error = e.message,
                    completedSteps = completedSteps,
                    compensatedSteps = compensated,
                )
            }
        }

        updateSagaStatus(sagaId, SagaStatus.COMPLETED)
        return SagaResult(
            sagaId = sagaId,
            status = SagaStatus.COMPLETED,
            context = context,
            completedSteps = completedSteps,
        )
    }

    /**
     * Retrieves all sagas in the dead letter queue.
     *
     * @return List of saga instances in dead letter state.
     */
    public suspend fun deadLetters(): List<SagaInstance> = newSuspendedTransaction(Dispatchers.IO, database) {
        val sql = """
            SELECT saga_id, saga_name, status, current_step, context, error, retry_count, created_at, updated_at
            FROM ${sagaConfig.tableName}
            WHERE status = 'DEAD_LETTER'
            ORDER BY updated_at DESC
        """.trimIndent()

        val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
        val stmt = conn.prepareStatement(sql, false)
        val rs = stmt.executeQuery().resultSet

        val results = mutableListOf<SagaInstance>()
        while (rs.next()) {
            results.add(
                SagaInstance(
                    sagaId = rs.getString("saga_id"),
                    sagaName = rs.getString("saga_name"),
                    status = SagaStatus.valueOf(rs.getString("status")),
                    currentStep = rs.getInt("current_step"),
                    context = rs.getString("context"),
                    error = rs.getString("error"),
                    retryCount = rs.getInt("retry_count"),
                    createdAt = Instant.parse(rs.getString("created_at")),
                    updatedAt = Instant.parse(rs.getString("updated_at")),
                ),
            )
        }
        results
    }

    private suspend fun compensate(
        sagaId: String,
        definition: SagaDefinition,
        completedSteps: List<String>,
        context: MutableMap<String, Any>,
    ): List<String> {
        val compensatedSteps = mutableListOf<String>()
        updateSagaStatus(sagaId, SagaStatus.COMPENSATING)

        // Compensate in reverse order
        for (stepName in completedSteps.reversed()) {
            val step = definition.steps.find { it.name == stepName } ?: continue
            try {
                logger.debug { "Compensating step '${step.name}' in saga '$sagaId'" }
                step.compensate(context)
                compensatedSteps.add(step.name)
            } catch (e: Exception) {
                logger.error(e) { "Compensation failed for step '${step.name}' in saga '$sagaId'" }
                metrics?.sagaCompensationFailed(definition.name, step.name)
                updateSagaStatus(sagaId, SagaStatus.FAILED)
                throw ConsensusError.CompensationFailed(sagaId, step.name, cause = e)
            }
        }

        updateSagaStatus(sagaId, SagaStatus.COMPENSATED)
        return compensatedSteps
    }

    private suspend fun persistSagaInstance(
        sagaId: String,
        sagaName: String,
        status: SagaStatus,
        currentStep: Int,
        context: Map<String, Any>,
    ) = newSuspendedTransaction(Dispatchers.IO, database) {
        val now = Instant.now().toString()
        val sql = """
            INSERT INTO ${sagaConfig.tableName} (saga_id, saga_name, status, current_step, context, retry_count, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, 0, ?, ?)
        """.trimIndent()

        val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
        val stmt = conn.prepareStatement(sql, false)
        stmt[1] = sagaId
        stmt[2] = sagaName
        stmt[3] = status.name
        stmt[4] = currentStep
        stmt[5] = "{}"
        stmt[6] = now
        stmt[7] = now
        stmt.executeUpdate()
    }

    private suspend fun updateSagaStatus(sagaId: String, status: SagaStatus) =
        newSuspendedTransaction(Dispatchers.IO, database) {
            val sql = "UPDATE ${sagaConfig.tableName} SET status = ?, updated_at = ? WHERE saga_id = ?"
            val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
            val stmt = conn.prepareStatement(sql, false)
            stmt[1] = status.name
            stmt[2] = Instant.now().toString()
            stmt[3] = sagaId
            stmt.executeUpdate()
        }

    private suspend fun updateSagaStep(sagaId: String, step: Int, status: SagaStatus) =
        newSuspendedTransaction(Dispatchers.IO, database) {
            val sql = "UPDATE ${sagaConfig.tableName} SET current_step = ?, status = ?, updated_at = ? WHERE saga_id = ?"
            val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
            val stmt = conn.prepareStatement(sql, false)
            stmt[1] = step
            stmt[2] = status.name
            stmt[3] = Instant.now().toString()
            stmt[4] = sagaId
            stmt.executeUpdate()
        }

    private suspend fun updateSagaRetry(sagaId: String, retryCount: Int, error: String?) =
        newSuspendedTransaction(Dispatchers.IO, database) {
            val sql = "UPDATE ${sagaConfig.tableName} SET retry_count = ?, error = ?, updated_at = ? WHERE saga_id = ?"
            val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
            val stmt = conn.prepareStatement(sql, false)
            stmt[1] = retryCount
            stmt[2] = error
            stmt[3] = Instant.now().toString()
            stmt[4] = sagaId
            stmt.executeUpdate()
        }

    private suspend fun moveToDeadLetter(sagaId: String, error: String?) =
        newSuspendedTransaction(Dispatchers.IO, database) {
            val sql = "UPDATE ${sagaConfig.tableName} SET status = 'DEAD_LETTER', error = ?, updated_at = ? WHERE saga_id = ?"
            val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
            val stmt = conn.prepareStatement(sql, false)
            stmt[1] = error
            stmt[2] = Instant.now().toString()
            stmt[3] = sagaId
            stmt.executeUpdate()
        }

    private suspend fun getRetryCount(sagaId: String): Int = newSuspendedTransaction(Dispatchers.IO, database) {
        val sql = "SELECT retry_count FROM ${sagaConfig.tableName} WHERE saga_id = ?"
        val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
        val stmt = conn.prepareStatement(sql, false)
        stmt[1] = sagaId
        val rs = stmt.executeQuery().resultSet
        if (rs.next()) rs.getInt("retry_count") else 0
    }

    private suspend fun getSagaInstance(sagaId: String): SagaInstance? = newSuspendedTransaction(Dispatchers.IO, database) {
        val sql = """
            SELECT saga_id, saga_name, status, current_step, context, error, retry_count, created_at, updated_at
            FROM ${sagaConfig.tableName}
            WHERE saga_id = ?
        """.trimIndent()

        val conn = org.jetbrains.exposed.sql.transactions.TransactionManager.current().connection
        val stmt = conn.prepareStatement(sql, false)
        stmt[1] = sagaId
        val rs = stmt.executeQuery().resultSet

        if (rs.next()) {
            SagaInstance(
                sagaId = rs.getString("saga_id"),
                sagaName = rs.getString("saga_name"),
                status = SagaStatus.valueOf(rs.getString("status")),
                currentStep = rs.getInt("current_step"),
                context = rs.getString("context"),
                error = rs.getString("error"),
                retryCount = rs.getInt("retry_count"),
                createdAt = Instant.parse(rs.getString("created_at")),
                updatedAt = Instant.parse(rs.getString("updated_at")),
            )
        } else {
            null
        }
    }
}

File: framework/backend/csharp/src/K1s0.Consensus/SagaOrchestrator.cs
using System.Text.Json;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Npgsql;

namespace K1s0.Consensus;

/// <summary>
/// Orchestrates saga execution with persistence, retry, and compensation support.
/// Uses PostgreSQL for durable saga state storage.
/// </summary>
public sealed class SagaOrchestrator
{
    private readonly SagaConfig _config;
    private readonly ILogger<SagaOrchestrator> _logger;
    private readonly string _connectionString;

    /// <summary>
    /// Creates a new <see cref="SagaOrchestrator"/>.
    /// </summary>
    /// <param name="options">Consensus configuration.</param>
    /// <param name="logger">Logger instance.</param>
    public SagaOrchestrator(IOptions<ConsensusConfig> options, ILogger<SagaOrchestrator> logger)
    {
        _config = options.Value.Saga;
        _logger = logger;
        _connectionString = ReadConnectionString(_config.ConnectionStringFile);
    }

    /// <summary>
    /// Executes a saga from the beginning.
    /// </summary>
    /// <typeparam name="TContext">The saga context type.</typeparam>
    /// <param name="definition">The saga definition.</param>
    /// <param name="context">The initial context.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The saga result.</returns>
    public async Task<SagaResult<TContext>> ExecuteAsync<TContext>(
        SagaDefinition<TContext> definition,
        TContext context,
        CancellationToken cancellationToken = default) where TContext : class
    {
        var sagaId = Guid.NewGuid().ToString("N");
        _logger.LogInformation("Starting saga {SagaName} with ID {SagaId}", definition.Name, sagaId);
        Metrics.SagaMetrics.StartedTotal.Inc();

        var instance = new SagaInstance
        {
            SagaId = sagaId,
            SagaName = definition.Name,
            Status = SagaStatus.Running,
            ContextJson = JsonSerializer.Serialize(context),
            CurrentStep = 0
        };

        await PersistInstanceAsync(instance, cancellationToken).ConfigureAwait(false);

        return await RunStepsAsync(definition, instance, context, 0, cancellationToken).ConfigureAwait(false);
    }

    /// <summary>
    /// Resumes a previously suspended or failed saga from the last completed step.
    /// </summary>
    /// <typeparam name="TContext">The saga context type.</typeparam>
    /// <param name="definition">The saga definition.</param>
    /// <param name="sagaId">The saga instance ID to resume.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The saga result.</returns>
    public async Task<SagaResult<TContext>> ResumeAsync<TContext>(
        SagaDefinition<TContext> definition,
        string sagaId,
        CancellationToken cancellationToken = default) where TContext : class
    {
        var instance = await LoadInstanceAsync(sagaId, cancellationToken).ConfigureAwait(false)
            ?? throw new ConsensusException("consensus.saga.not_found", $"Saga instance '{sagaId}' not found.");

        var context = instance.DeserializeContext<TContext>()
            ?? throw new ConsensusException("consensus.saga.invalid_context", $"Failed to deserialize context for saga '{sagaId}'.");

        _logger.LogInformation("Resuming saga {SagaId} from step {Step}", sagaId, instance.CurrentStep);

        return await RunStepsAsync(definition, instance, context, instance.CurrentStep, cancellationToken).ConfigureAwait(false);
    }

    /// <summary>
    /// Returns all saga instances that have reached the dead letter state (failed compensation).
    /// </summary>
    /// <param name="sagaName">Optional filter by saga name.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Dead letter saga instances.</returns>
    public async Task<IReadOnlyList<SagaInstance>> DeadLettersAsync(string? sagaName = null, CancellationToken cancellationToken = default)
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync(cancellationToken).ConfigureAwait(false);

        var sql = sagaName is not null
            ? $"SELECT saga_id, saga_name, status, context_json, current_step, created_at, updated_at, errors_json FROM {_config.DeadLetterTableName} WHERE saga_name = @name ORDER BY created_at DESC"
            : $"SELECT saga_id, saga_name, status, context_json, current_step, created_at, updated_at, errors_json FROM {_config.DeadLetterTableName} ORDER BY created_at DESC";

        await using var cmd = new NpgsqlCommand(sql, conn);
        if (sagaName is not null)
        {
            cmd.Parameters.AddWithValue("name", sagaName);
        }

        var results = new List<SagaInstance>();
        await using var reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
        while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            results.Add(ReadInstance(reader));
        }

        return results.AsReadOnly();
    }

    private async Task<SagaResult<TContext>> RunStepsAsync<TContext>(
        SagaDefinition<TContext> definition,
        SagaInstance instance,
        TContext context,
        int startStep,
        CancellationToken cancellationToken) where TContext : class
    {
        var errors = new Dictionary<string, string>();
        var completedSteps = new List<int>();

        for (var i = startStep; i < definition.Steps.Count; i++)
        {
            var step = definition.Steps[i];
            var retryPolicy = definition.GetRetryPolicy(step.Name);

            instance.CurrentStep = i;
            instance.ContextJson = JsonSerializer.Serialize(context);
            await UpdateInstanceAsync(instance, cancellationToken).ConfigureAwait(false);

            var success = false;

            for (var attempt = 0; attempt <= retryPolicy.MaxRetries; attempt++)
            {
                try
                {
                    using var stepCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                    stepCts.CancelAfter(_config.DefaultStepTimeout);

                    context = await step.ExecuteAsync(context, stepCts.Token).ConfigureAwait(false);
                    completedSteps.Add(i);
                    success = true;
                    _logger.LogDebug("Saga {SagaId} step {StepName} completed", instance.SagaId, step.Name);
                    Metrics.SagaMetrics.StepsCompletedTotal.Inc();
                    break;
                }
                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
                {
                    throw;
                }
                catch (Exception ex)
                {
                    _logger.LogWarning(ex, "Saga {SagaId} step {StepName} attempt {Attempt} failed",
                        instance.SagaId, step.Name, attempt + 1);

                    if (attempt < retryPolicy.MaxRetries)
                    {
                        var delay = retryPolicy.GetDelay(attempt);
                        await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
                    }
                    else
                    {
                        errors[step.Name] = ex.Message;
                        Metrics.SagaMetrics.StepsFailedTotal.Inc();
                    }
                }
            }

            if (!success)
            {
                // Compensate completed steps in reverse
                _logger.LogWarning("Saga {SagaId} compensating from step {Step}", instance.SagaId, i);
                return await CompensateAsync(definition, instance, context, completedSteps, errors, cancellationToken).ConfigureAwait(false);
            }
        }

        instance.Status = SagaStatus.Completed;
        instance.ContextJson = JsonSerializer.Serialize(context);
        instance.UpdatedAt = DateTimeOffset.UtcNow;
        await UpdateInstanceAsync(instance, cancellationToken).ConfigureAwait(false);

        _logger.LogInformation("Saga {SagaId} completed successfully", instance.SagaId);
        Metrics.SagaMetrics.CompletedTotal.Inc();

        return new SagaResult<TContext>
        {
            Status = SagaStatus.Completed,
            Context = context,
            SagaId = instance.SagaId,
            LastCompletedStep = definition.Steps.Count - 1
        };
    }

    private async Task<SagaResult<TContext>> CompensateAsync<TContext>(
        SagaDefinition<TContext> definition,
        SagaInstance instance,
        TContext context,
        List<int> completedSteps,
        Dictionary<string, string> errors,
        CancellationToken cancellationToken) where TContext : class
    {
        var compensationFailed = false;

        for (var i = completedSteps.Count - 1; i >= 0; i--)
        {
            var stepIndex = completedSteps[i];
            var step = definition.Steps[stepIndex];

            try
            {
                context = await step.CompensateAsync(context, cancellationToken).ConfigureAwait(false);
                _logger.LogDebug("Saga {SagaId} compensated step {StepName}", instance.SagaId, step.Name);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Saga {SagaId} compensation failed for step {StepName}", instance.SagaId, step.Name);
                errors[$"{step.Name}_compensation"] = ex.Message;
                compensationFailed = true;
            }
        }

        var status = compensationFailed ? SagaStatus.Failed : SagaStatus.Compensated;

        instance.Status = status;
        instance.ContextJson = JsonSerializer.Serialize(context);
        instance.ErrorsJson = JsonSerializer.Serialize(errors);
        instance.UpdatedAt = DateTimeOffset.UtcNow;
        await UpdateInstanceAsync(instance, cancellationToken).ConfigureAwait(false);

        if (compensationFailed)
        {
            _logger.LogError("Saga {SagaId} moved to dead letter", instance.SagaId);
            await MoveToDeadLetterAsync(instance, cancellationToken).ConfigureAwait(false);
            Metrics.SagaMetrics.DeadLetterTotal.Inc();
        }
        else
        {
            Metrics.SagaMetrics.CompensatedTotal.Inc();
        }

        return new SagaResult<TContext>
        {
            Status = status,
            Context = context,
            SagaId = instance.SagaId,
            Errors = errors.AsReadOnly(),
            LastCompletedStep = completedSteps.Count > 0 ? completedSteps[^1] : -1
        };
    }

    private async Task PersistInstanceAsync(SagaInstance instance, CancellationToken cancellationToken)
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync(cancellationToken).ConfigureAwait(false);

        var sql = $@"
            INSERT INTO {_config.TableName} (saga_id, saga_name, status, context_json, current_step, created_at, updated_at, errors_json)
            VALUES (@id, @name, @status, @context::jsonb, @step, @created, @updated, @errors::jsonb)";

        await using var cmd = new NpgsqlCommand(sql, conn);
        cmd.Parameters.AddWithValue("id", instance.SagaId);
        cmd.Parameters.AddWithValue("name", instance.SagaName);
        cmd.Parameters.AddWithValue("status", (int)instance.Status);
        cmd.Parameters.AddWithValue("context", instance.ContextJson);
        cmd.Parameters.AddWithValue("step", instance.CurrentStep);
        cmd.Parameters.AddWithValue("created", instance.CreatedAt);
        cmd.Parameters.AddWithValue("updated", instance.UpdatedAt);
        cmd.Parameters.AddWithValue("errors", instance.ErrorsJson);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    private async Task UpdateInstanceAsync(SagaInstance instance, CancellationToken cancellationToken)
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync(cancellationToken).ConfigureAwait(false);

        var sql = $@"
            UPDATE {_config.TableName}
            SET status = @status, context_json = @context::jsonb, current_step = @step, updated_at = @updated, errors_json = @errors::jsonb
            WHERE saga_id = @id";

        await using var cmd = new NpgsqlCommand(sql, conn);
        cmd.Parameters.AddWithValue("id", instance.SagaId);
        cmd.Parameters.AddWithValue("status", (int)instance.Status);
        cmd.Parameters.AddWithValue("context", instance.ContextJson);
        cmd.Parameters.AddWithValue("step", instance.CurrentStep);
        cmd.Parameters.AddWithValue("updated", DateTimeOffset.UtcNow);
        cmd.Parameters.AddWithValue("errors", instance.ErrorsJson);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    private async Task<SagaInstance?> LoadInstanceAsync(string sagaId, CancellationToken cancellationToken)
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync(cancellationToken).ConfigureAwait(false);

        var sql = $"SELECT saga_id, saga_name, status, context_json, current_step, created_at, updated_at, errors_json FROM {_config.TableName} WHERE saga_id = @id";

        await using var cmd = new NpgsqlCommand(sql, conn);
        cmd.Parameters.AddWithValue("id", sagaId);

        await using var reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
        if (!await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            return null;
        }

        return ReadInstance(reader);
    }

    private async Task MoveToDeadLetterAsync(SagaInstance instance, CancellationToken cancellationToken)
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync(cancellationToken).ConfigureAwait(false);

        var sql = $@"
            INSERT INTO {_config.DeadLetterTableName} (saga_id, saga_name, status, context_json, current_step, created_at, updated_at, errors_json)
            VALUES (@id, @name, @status, @context::jsonb, @step, @created, @updated, @errors::jsonb)
            ON CONFLICT (saga_id) DO UPDATE SET status = @status, context_json = @context::jsonb, updated_at = @updated, errors_json = @errors::jsonb";

        await using var cmd = new NpgsqlCommand(sql, conn);
        cmd.Parameters.AddWithValue("id", instance.SagaId);
        cmd.Parameters.AddWithValue("name", instance.SagaName);
        cmd.Parameters.AddWithValue("status", (int)instance.Status);
        cmd.Parameters.AddWithValue("context", instance.ContextJson);
        cmd.Parameters.AddWithValue("step", instance.CurrentStep);
        cmd.Parameters.AddWithValue("created", instance.CreatedAt);
        cmd.Parameters.AddWithValue("updated", instance.UpdatedAt);
        cmd.Parameters.AddWithValue("errors", instance.ErrorsJson);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    private static SagaInstance ReadInstance(NpgsqlDataReader reader) => new()
    {
        SagaId = reader.GetString(0),
        SagaName = reader.GetString(1),
        Status = (SagaStatus)reader.GetInt32(2),
        ContextJson = reader.GetString(3),
        CurrentStep = reader.GetInt32(4),
        CreatedAt = reader.GetFieldValue<DateTimeOffset>(5),
        UpdatedAt = reader.GetFieldValue<DateTimeOffset>(6),
        ErrorsJson = reader.GetString(7)
    };

    private static string ReadConnectionString(string filePath)
    {
        if (string.IsNullOrWhiteSpace(filePath))
        {
            throw new ConsensusException(
                "consensus.config.missing_connection_string",
                "Saga connection_string_file is not configured.");
        }

        return File.ReadAllText(filePath).Trim();
    }
}
